conans/client/cache/remote_registry.py

```diff
@@ -40,316 +36,307 @@ def load_registry_txt(contents):
-           "package_references": prefs}
-
-    return json.dumps(ret, indent=True)
-
-
-def load_registry(contents):
+def load_old_registry_json(contents):
     """From json"""
     data = json.loads(contents)
-    remotes = OrderedDict()
+    remotes = Remotes()
     refs = data.get("references", {})
     prefs = data.get("package_references", {})
     for r in data["remotes"]:
-        remotes[r["name"]] = (r["url"], r["verify_ssl"])
+        remotes.add(r["name"], r["url"], r["verify_ssl"])
     return remotes, refs, prefs
 
 
-def migrate_registry_file(path, new_path):
-    try:
-        remotes, refs = load_registry_txt(load(path))
-        save(new_path, dump_registry(remotes, refs, {}))
-    except Exception as e:
-        raise ConanException("Cannot migrate registry.txt to registry.json: %s" % str(e))
-    else:
-        os.unlink(path)
+def migrate_registry_file(cache, out):
+    REGISTRY = "registry.txt"
+    REGISTRY_JSON = "registry.json"
+    folder = cache.conan_folder
+    reg_json_path = os.path.join(folder, REGISTRY_JSON)
+    reg_txt_path = os.path.join(folder, REGISTRY)
 
+    def add_ref_remote(reference, remotes, remote_name):
+        ref = ConanFileReference.loads(reference, validate=True)
+        remote = remotes[remote_name]
+        with cache.package_layout(ref).update_metadata() as metadata:
+            metadata.recipe.remote = remote.name
 
-class _Registry(object):
+    def add_pref_remote(pkg_ref, remotes, remote_name):
+        pref = PackageReference.loads(pkg_ref, validate=True)
+        remote = remotes[remote_name]
+        with cache.package_layout(pref.ref).update_metadata() as metadata:
+            metadata.packages[pref.id].remote = remote.name
 
-    def __init__(self, filename, lockfile, output):
-        self._filename = filename
-        self._lockfile = lockfile
-        self._output = output
+    try:
+        if os.path.exists(reg_json_path):
+            out.warn("%s has been deprecated. Migrating to remotes.json" % REGISTRY)
+            remotes, refs, prefs = load_old_registry_json(load(reg_json_path))
+            remotes.save(cache.registry_path)
+            for ref, remote_name in refs.items():
+                add_ref_remote(ref, remotes, remote_name)
+            for pref, remote_name in prefs.items():
+                add_pref_remote(pref, remotes, remote_name)
+            os.remove(reg_json_path)
+        elif os.path.exists(reg_txt_path):
+            out.warn("%s has been deprecated. Migrating to remotes.json" % REGISTRY_JSON)
+            remotes, refs = load_registry_txt(load(reg_txt_path))
+            remotes.save(cache.registry_path)
+            for ref, remote_name in refs.items():
+                add_ref_remote(ref, remotes, remote_name)
+            os.remove(reg_txt_path)
 
-    def _load(self):
-        tmp = load(self._filename)
-        return load_registry(tmp)
+    except Exception as e:
+        raise ConanException("Cannot migrate old registry: %s" % str(e))
```

Comment:
If it fails once, it will always fail. If I were a user I would get stuck in the "Cannot migrate old registry: <any error>" message. Is there any action I can do as a user to solve it without opening an issue asking for help?

Reply:
Yes, I have added "remove the registry file and try again". That might solve most of the migrations issues, but might lose some data. I'd prefer users reporting issues so we can correctly fix it.